net: lwm2m: update client tx timestamp before sending message

marinjurjevic opened the pull request:
TX timestamp tracking mechanisms predates async socket io introduction.
Timestamp tracking is needed for notifying client when RX can be
shut down. Since the check is performed inside registration done
state, TX timestamp is invalid so we update it directly.

Signed-off-by: Marin Jurjević <marin.jurjevic@hotmail.com>

marinjurjevic commented:
Tracking last TX timestamp logic should probably be refactored. This was fine before moving to async socket sending https://github.com/zephyrproject-rtos/zephyr/commit/32989a38f0c8911e83d64fc898e19d29b0800f75. 

The problem is that condition for firing RX_OFF event is checked right after sending registration/update and it relies on `sm_send_registration` updating this value.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -920,6 +920,13 @@ static int sm_registration_done(void)
+		 * when performing check should ENGINE_REGISTRATION_DONE_RX_OFF
+		 * be generated below.
+		 */
+		engine_update_tx_time();
```
I wonder, maybe it'd be better to move `engine_update_tx_time()` into `lwm2m_send_message_async()` instead of covering corner cases in lwm2m_rd_client? I am aware that it's not the actual timestamp of putting the packet on the wire, but the delay between queueing the packet and the actual send call should be negligible, given the timeout value (which is 93 seconds by default). What do you think?

marinjurjevic replied:
> but the delay between queueing the packet and the actual send call should be negligible,

Sounds good. Initially I did not want to put it there (`lwm2m_send_message_async`), but in the function where the bug actually occurs so fix is more obvious. But since the delay really is negligible I'll move it there.

marinjurjevic committed:
net: lwm2m: update client tx timestamp before sending message

Update client tx timestamp right after message is added to list
of outgoing messages. Delay between when message is generated and
sent is negligible. This will prevents bugs that appear when using
queue mode, where internal engine logic depends on timestamp being
updated when message is generated.

Signed-off-by: Marin Jurjević <marin.jurjevic@hotmail.com>
